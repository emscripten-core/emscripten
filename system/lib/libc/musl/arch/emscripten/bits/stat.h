// XXX Emscripten in sync with musl/arch/i386/bits/stat.h except for the padding and 64-bit time_t redirections change.

/* copied from kernel definition, but with padding replaced
 * by the corresponding correctly-sized userspace types. */
struct stat {
	dev_t st_dev;
#ifndef __EMSCRIPTEN__
	int __st_dev_padding;
	long __st_ino_truncated;
#endif
	mode_t st_mode;
	nlink_t st_nlink;
	uid_t st_uid;
	gid_t st_gid;
	dev_t st_rdev;
#ifndef __EMSCRIPTEN__
	int __st_rdev_padding;
#endif
	off_t st_size;
	blksize_t st_blksize;
	blkcnt_t st_blocks;
#ifndef __EMSCRIPTEN__ // XXX Emscripten no need to activate the symbol redirections for 64-bit time_t.
	struct {
		long tv_sec;
		long tv_nsec;
	} __st_atim32, __st_mtim32, __st_ctim32;
#endif
	struct timespec st_atim;
	struct timespec st_mtim;
	struct timespec st_ctim;
	ino_t st_ino;
};
