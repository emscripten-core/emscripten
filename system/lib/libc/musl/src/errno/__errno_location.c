#include <errno.h>
#include "pthread_impl.h"

#if __EMSCRIPTEN__
// For emscripten we use TLS here instead of `__pthread_self`, so that in single
// threaded builds this gets lowered away to normal global variable.
static _Thread_local int __errno_storage = 0;
#endif

int *__errno_location(void)
{
#if __EMSCRIPTEN__
	return &__errno_storage;
#else
	return &__pthread_self()->errno_val;
#endif
}

weak_alias(__errno_location, ___errno_location);
