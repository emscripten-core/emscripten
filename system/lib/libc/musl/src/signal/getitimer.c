#include <sys/time.h>
#include "syscall.h"

#ifdef __EMSCRIPTEN__
#include <emscripten/emscripten.h>
void __getitimer(int which, struct itimerval *old, double now);
#endif

int getitimer(int which, struct itimerval *old)
{
#ifdef __EMSCRIPTEN__
	if (which > ITIMER_PROF) return EINVAL;
	__getitimer(which, old, emscripten_get_now());
	return 0;
#else
	if (sizeof(time_t) > sizeof(long)) {
		long old32[4];
		int r = __syscall(SYS_getitimer, which, old32);
		if (!r) {
			old->it_interval.tv_sec = old32[0];
			old->it_interval.tv_usec = old32[1];
			old->it_value.tv_sec = old32[2];
			old->it_value.tv_usec = old32[3];
		}
		return __syscall_ret(r);
	}
	return syscall(SYS_getitimer, which, old);
#endif
}
