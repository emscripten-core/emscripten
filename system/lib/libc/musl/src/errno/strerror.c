#include <errno.h>
#include <stddef.h>
#include <string.h>
#include "locale_impl.h"

/* mips has one error code outside of the 8-bit range due to a
 * historical typo, so we just remap it. */
#if EDQUOT==1133
#define EDQUOT_ORIG 1133
#undef  EDQUOT
#define EDQUOT 109
#endif

static const struct errmsgstr_t {
#define E(n, s) char str##n[sizeof(s)];
#include "__strerror.h"
#undef E
} errmsgstr = {
#define E(n, s) s,
#include "__strerror.h"
#undef E
};

static const unsigned short errmsgidx[] = {
#define E(n, s) [n] = offsetof(struct errmsgstr_t, str##n),
#include "__strerror.h"
#undef E
};

char *__strerror_l(int e, locale_t loc)
{
	const char *s;
#ifdef EDQUOT_ORIG
	if (e==EDQUOT) e=0;
	else if (e==EDQUOT_ORIG) e=EDQUOT;
#endif
	if (e >= sizeof errmsgidx / sizeof *errmsgidx) e = 0;
	s = (char *)&errmsgstr + errmsgidx[e];
#ifdef __EMSCRIPTEN__
	// strerror is a (debug) dependency of many emscripten syscalls which mean it
	// must be excluded from LTO, along with all of its dependencies.
	// In order to limit the transitive dependencies we disable localization of
	// rrno messages here.
	return (char *)s;
#else
	return (char *)LCTRANS(s, LC_MESSAGES, loc);
#endif
}

char *strerror(int e)
{
#ifdef __EMSCRIPTEN__
	return __strerror_l(e, NULL);
#else
	return __strerror_l(e, CURRENT_LOCALE);
#endif
}

weak_alias(__strerror_l, strerror_l);
