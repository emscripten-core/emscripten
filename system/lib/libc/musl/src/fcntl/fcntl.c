#define _GNU_SOURCE
#include <fcntl.h>
#include <stdarg.h>
#include <errno.h>
#include "syscall.h"

#ifdef __EMSCRIPTEN__
__attribute__((no_sanitize("address")))
#endif
int fcntl(int fd, int cmd, ...)
{
	unsigned long arg;
	// XXX Emscripten: According to the va_arg man page it is undefined behaviour to
	// read arguments that are not passed.  This can lead to a false positive
	// in SAFE_HEAP, so avoid it.
	if (cmd != F_GETFL && cmd != F_GETFD && cmd != F_GETOWN) {
		va_list ap;
		va_start(ap, cmd);
		arg = va_arg(ap, unsigned long);
		va_end(ap);
	}
	if (cmd == F_SETFL) arg |= O_LARGEFILE;
	if (cmd == F_SETLKW) return syscall_cp(SYS_fcntl, fd, cmd, (void *)arg);
	if (cmd == F_GETOWN) {
		struct f_owner_ex ex;
		int ret = __syscall(SYS_fcntl, fd, F_GETOWN_EX, &ex);
#ifdef __EMSCRIPTEN__
		// XXX Emscripten: Mirror the behaviour/limitation of glibc which
		// will misinterpret negative PIDs in range of -1 to -4095 as being
		// errno values.
		if (ret == -EINVAL) ret = __syscall(SYS_fcntl, fd, cmd, (void *)arg);
#else
		if (ret == -EINVAL) return __syscall(SYS_fcntl, fd, cmd, (void *)arg);
#endif
		if (ret) return __syscall_ret(ret);
		return ex.type == F_OWNER_PGRP ? -ex.pid : ex.pid;
	}
	if (cmd == F_DUPFD_CLOEXEC) {
		int ret = __syscall(SYS_fcntl, fd, F_DUPFD_CLOEXEC, arg);
		if (ret != -EINVAL) {
#ifndef __EMSCRIPTEN__ // CLOEXEC makes no sense for a single process
			if (ret >= 0) __syscall(SYS_fcntl, ret, F_SETFD, FD_CLOEXEC);
#endif
			return __syscall_ret(ret);
		}
		ret = __syscall(SYS_fcntl, fd, F_DUPFD_CLOEXEC, 0);
		if (ret != -EINVAL) {
#ifdef __EMSCRIPTEN__
			if (ret >= 0) __wasi_fd_close(ret);
#else
			if (ret >= 0) __syscall(SYS_close, ret);
#endif
			return __syscall_ret(-EINVAL);
		}
		ret = __syscall(SYS_fcntl, fd, F_DUPFD, arg);
#ifndef __EMSCRIPTEN__ // CLOEXEC makes no sense for a single process
		if (ret >= 0) __syscall(SYS_fcntl, ret, F_SETFD, FD_CLOEXEC);
#endif
		return __syscall_ret(ret);
	}
	switch (cmd) {
	case F_SETLK:
	case F_GETLK:
	case F_GETOWN_EX:
	case F_SETOWN_EX:
		return syscall(SYS_fcntl, fd, cmd, (void *)arg);
	default:
		return syscall(SYS_fcntl, fd, cmd, arg);
	}
}
