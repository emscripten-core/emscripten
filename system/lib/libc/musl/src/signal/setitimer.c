#include <sys/time.h>
#include <errno.h>
#include "syscall.h"

#define IS32BIT(x) !((x)+0x80000000ULL>>32)

#ifdef __EMSCRIPTEN__
#include <emscripten/emscripten.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>

#include "emscripten_internal.h"

// Timeouts can either fire directly from the JS event loop (which calls
// `_emscripten_timeout`), or from `_emscripten_check_timers` (which is called
// from `_emscripten_yield`).  In order to be able to check the timers here we
// cache the current timeout and interval for each the 3 types of timer
// (ITIMER_PROF/ITIMER_VIRTUAL/ITIMER_REAL).
static double current_timeout_ms[3];
static double current_intervals_ms[3];

#define MAX(a,b) ((a)>(b)?(a):(b))

void __getitimer(int which, struct itimerval *old, double now)
{
	double remaining_ms = MAX(current_timeout_ms[which] - now, 0);
	old->it_value.tv_sec = remaining_ms / 1000;
	old->it_value.tv_usec = remaining_ms * 1000;
	old->it_interval.tv_sec = current_intervals_ms[which] / 1000;
	old->it_interval.tv_usec = current_intervals_ms[which] * 1000;
}

void _emscripten_timeout(int which, double now)
{
	int signum = SIGALRM;
	if (which == ITIMER_PROF)
		signum = SIGPROF;
	else if (which == ITIMER_VIRTUAL)
		signum = SIGVTALRM;
	double next_timeout = 0.0;
	if (current_intervals_ms[which]) {
		// If time went backwards, schedule the next timer as if it didn't.
		now = __builtin_wasm_max_f64(now, current_timeout_ms[which]);
		// The next alarm is due 'interval' ms after the previous one.
		// If this alarm was delayed, that is sooner than 'interval' ms
		// from now. The delay could even be so long that we missed the
		// next alarm(s) entirely. Schedule the alarm for the next
		// multiple of 'interval' ms from the original due time.
		uint64_t intervals =
			(uint64_t)(now - current_timeout_ms[which]) /
			  (uint64_t)current_intervals_ms[which] +
			1;
		current_timeout_ms[which] +=
			intervals * current_intervals_ms[which];
		next_timeout = current_timeout_ms[which] - now;
	} else {
		current_timeout_ms[which] = 0;
	}
	_setitimer_js(which, next_timeout);
	raise(signum);
}

void _emscripten_check_timers(double now)
{
	for (int which = 0; which < 3; which++) {
		if (current_timeout_ms[which]) {
			// Only call out to JS to get the current time if it was not passed in
			// *and* we have one or more timers set.
			if (!now)
			 	now = emscripten_get_now();
			if (now >= current_timeout_ms[which])
				_emscripten_timeout(which, now);
		}
	}
}
#endif

int setitimer(int which, const struct itimerval *restrict new, struct itimerval *restrict old)
{
#ifdef __EMSCRIPTEN__
	if (which > ITIMER_PROF) return EINVAL;
	double now = emscripten_get_now();
	if (old) {
		__getitimer(which, old, now);
	}
	if (new->it_value.tv_sec || new->it_value.tv_usec) {
		current_timeout_ms[which] = now + new->it_value.tv_sec * 1000 + new->it_value.tv_usec / 1000;
		current_intervals_ms[which] = new->it_interval.tv_sec * 1000 + new->it_interval.tv_usec / 1000;
	} else {
		current_timeout_ms[which] = 0;
		current_intervals_ms[which] = 0;
	}
	return _setitimer_js(which, new->it_value.tv_sec * 1000 + new->it_value.tv_usec / 1000);
#else
	if (sizeof(time_t) > sizeof(long)) {
		time_t is = new->it_interval.tv_sec, vs = new->it_value.tv_sec;
		long ius = new->it_interval.tv_usec, vus = new->it_value.tv_usec;
		if (!IS32BIT(is) || !IS32BIT(vs))
			return __syscall_ret(-ENOTSUP);
		long old32[4];
		int r = __syscall(SYS_setitimer, which,
			((long[]){is, ius, vs, vus}), old32);
		if (!r && old) {
			old->it_interval.tv_sec = old32[0];
			old->it_interval.tv_usec = old32[1];
			old->it_value.tv_sec = old32[2];
			old->it_value.tv_usec = old32[3];
		}
		return __syscall_ret(r);
	}
	return syscall(SYS_setitimer, which, new, old);
#endif
}
