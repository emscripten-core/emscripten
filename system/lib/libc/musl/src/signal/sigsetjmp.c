#if __EMSCRIPTEN__
#include <setjmp.h>
#include <signal.h>

int sigsetjmp(jmp_buf sigjmp_buf, int savesigs) {
    return setjmp(sigjmp_buf);
}
#endif
