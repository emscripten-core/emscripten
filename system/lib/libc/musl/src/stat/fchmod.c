#ifdef __EMSCRIPTEN__
#include <stropts.h>
#endif
#include <sys/stat.h>
#include <errno.h>
#include <fcntl.h>
#include "syscall.h"

int fchmod(int fd, mode_t mode)
{
	int ret = __syscall(SYS_fchmod, fd, mode);
#if __EMSCRIPTEN__
	if (ret != -EBADF || !__wasi_fd_is_valid(fd))
		return __syscall_ret(ret);
#else
	if (ret != -EBADF || __syscall(SYS_fcntl, fd, F_GETFD) < 0)
		return __syscall_ret(ret);
#endif

	char buf[15+3*sizeof(int)];
	__procfdname(buf, fd);
#ifdef SYS_chmod
	return syscall(SYS_chmod, buf, mode);
#else
	return syscall(SYS_fchmodat, AT_FDCWD, buf, mode);
#endif
}
