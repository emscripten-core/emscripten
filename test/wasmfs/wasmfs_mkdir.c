/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

// FIXME: Merge with unlink test

int main() {
  // Try to make a directory under the root directory.
  errno = 0;
  mkdir("/working", 0777);
  mkdir("/foobar", 01777);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == 0);

  // Check that the file type is correct on mode (0777 = S_IRWXUGO)
  int fdStat = open("/working", O_RDONLY | O_DIRECTORY);
  struct stat directory;
  fstat(fdStat, &directory);
  assert((directory.st_mode & S_IFMT) == S_IFDIR);
  assert(directory.st_mode == (S_IRWXUGO | S_IFDIR));
  close(fdStat);

  // Check that the file type is correct on mode (01777 = S_ISVTX | S_IRWXUGO)
  int fdStat2 = open("/foobar", O_RDONLY | O_DIRECTORY);
  struct stat directory2;
  fstat(fdStat2, &directory2);
  assert(directory2.st_mode == (S_IRWXUGO | S_ISVTX | S_IFDIR));
  close(fdStat2);

  // Try to create a file in the same directory.
  int fd = open("/working/test", O_RDWR | O_CREAT, 0777);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == 0);

  // Try to read and write to the same file.
  const char* msg = "Test\n";
  errno = 0;
  write(fd, msg, strlen(msg));
  assert(errno == 0);
  lseek(fd, 0, SEEK_SET);
  char buf[100] = {};
  errno = 0;
  read(fd, buf, sizeof(buf));
  assert(errno == 0);
  printf("%s", buf);
  close(fd);

  // In Linux and WasmFS, an empty pathname returns ENOENT.
  errno = 0;
  mkdir("", 0777);
  assert(errno == ENOENT);

  // Try to make the root directory.
  errno = 0;
  mkdir("/", 0777);
  assert(errno == EEXIST);

  // Try to make a directory that exists already.
  errno = 0;
  mkdir("/dev", 0777);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == EEXIST);

  // Try to make a directory with a path component that is not a directory.
  errno = 0;
  mkdir("/dev/stdout/fake-directory", 0777);
  assert(errno == ENOTDIR);

  // Try to make a directory with a path component that does not exist.
  errno = 0;
  mkdir("/dev/false-path/fake-directory", 0777);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == ENOENT);

  // Try to make a directory under the `working` directory.
  errno = 0;
  mkdir("/working/new-directory", 0777);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == 0);

  // Try to make a directory with a name that is longer than WASMFS_NAME_MAX.
  // In Linux, creating a directory with a name that is longer than 255
  // characters returns ENAMETOOLONG.
  errno = 0;
  mkdir("/working/"
        "0000000001000000000200000000030000000004000000000500000000060000000007"
        "0000000008000000000900000000000000000001000000000200000000030000000004"
        "0000000005000000000600000000070000000008000000000900000000000000000001"
        "0000000002000000000300000000040000000005123456",
        0777);
#ifdef WASMFS
  assert(errno == ENAMETOOLONG);
#endif

  return 0;
}
