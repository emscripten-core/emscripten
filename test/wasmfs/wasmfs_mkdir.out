Errno: No error information
Errno: No error information
Test
Errno: File exists
Errno: No such file or directory
Errno: No error information
